src/sync/ms_queue.rs

```diff
@@ -0,0 +1,423 @@
+                .map(|new| {
+                    // try to move the tail pointer forward
+                    let _ = self.tail.compare_and_set(onto, new, Release, scope);
+                })
+                .map_err(|(_, new)| new)
+        }
+    }
+
+    /// Add `t` to the back of the queue, possibly waking up threads
+    /// blocked on `pop`.
+    pub fn push(&self, t: T, scope: &Scope) {
+        let mut new = Owned::new(Node {
+            data: t,
+            next: Atomic::null(),
+        });
+
+        loop {
+            // We push onto the tail, so we'll start optimistically by looking
+            // there first.
+            let tail = self.tail.load(Acquire, scope);
+
+            // Attempt to push onto the `tail` snapshot; fails if
+            // `tail.next` has changed, which will always be the case if the
+            // queue has transitioned to blocking mode.
+            match self.push_internal(tail, new, scope) {
+                Ok(_) => break,
+                Err(temp) => {
+                    // retry
+                    new = temp
+                }
+            }
+        }
+    }
+
+    #[inline(always)]
+    // Attempt to pop a data node. `Ok(None)` if queue is empty or in blocking
+    // mode; `Err(())` if lost race to pop.
+    fn pop_internal(&self, scope: &Scope) -> Result<Option<T>, ()> {
+        let head = self.head.load(Acquire, scope);
+        let h = unsafe { head.deref() };
+        let next = h.next.load(Acquire, scope);
+        match unsafe { next.as_ref() } {
+            None => Ok(None),
+            Some(n) => unsafe {
+                if self.head
+                    .compare_and_set(head, next, Release, scope)
+                    .is_ok()
+                {
+                    scope.defer_free(head);
+                    Ok(Some(ptr::read(&n.data)))
+                } else {
+                    Err(())
+                }
+            },
+        }
+    }
+
+    /// Check if this queue is empty.
+    pub fn is_empty(&self) -> bool {
+        pin(|scope| {
+            let head = self.head.load(Acquire, scope);
+            let h = unsafe { head.deref() };
+            h.next.load(Acquire, scope).is_null()
+        })
+    }
+
+    /// Attempt to dequeue from the front.
+    ///
+    /// Returns `None` if the queue is observed to be empty.
+    pub fn try_pop(&self, scope: &Scope) -> Option<T> {
+        loop {
+            if let Ok(r) = self.pop_internal(scope) {
+                return r;
+            }
+        }
+    }
+
+    /// Attempt to dequeue from the front, if the item satisfies the given condition.
+    ///
+    /// Returns `None` if the queue is observed to be empty, or the head does not satisfy the given
+    /// condition.
+    pub fn try_pop_if<F>(&self, condition: F, scope: &Scope) -> Option<T>
+    where
+        F: Fn(&T) -> bool,
+    {
+        loop {
+            if let Ok(head) = self.pop_internal(scope) {
```

Comment:
Hm, I don't see how this implementation can get into an unsafe situation given that it evaluates the `condition` _after_ the element is popped and is thus almost equivalent to `let a = try_pop(); if cond(a) { return a; }`. Granted, this behaviour does not actually meet the specification and is more like `try_pop_and_then_return_if`.